### PR TITLE
fix #35305, need escaping when printing string macro calls

### LINKED ISSUE #35305: String representation of expressions with quotes in non-standard string literals
The string representation of an expression that contains quotes `"` in triple-quoted non-standard string literals (or escaped quotes in quoted non-standard string literals) is not faithfully any more. The resulting text does not represent valid Julia code and cannot be `include`d again. in Julia v1.4

```
julia> print(:(r = r"""45 " """))
r = r"45 " "
```

```
julia> print(:(r = r"45 \" "))
r = r"45 " "
```

This does not happen for "normal" strings
```
julia> print(:(r = "45 \" "))
r = "45 \" "
```

and also used to work in Julia v1.3.1. See also https://discourse.julialang.org/t/string-representation-of-expression/36729/3?u=pfarndt

### PR DESCRIPTION
**Pull request by JeffBezanson**
fix #35305

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]
I don't think this actually fixes it, because of https://github.com/JuliaLang/julia/issues/29580, but it is better.

**Comment by JeffBezanson** [comment]
It looks like that is about printing `Regex` and `SubstitutionString`; this is for expressions:
```
julia> :(r"\\\\")
:(r"\\\\")
```
But I should change this to only escape backslashes and quotes I think.

**Comment by vtjnash** [comment]
True, but those happen to be `raw""` strings also, like these.

Also, some part of that fix is included in https://github.com/JuliaLang/julia/pull/34111/files#diff-85598a9f3fb3fcced8c43d0472c30888R545

**Comment by JeffBezanson** [comment]
Oh, nice, we should add that code and use it here.

**Comment by JeffBezanson** [comment]
Added the commit from #34111. I edited it a bit to not print surrounding `"`s, since `escape_string` doesn't.

**Comment by mgkuhn** [comment]
@JeffBezanson The reason I printed the surrounding `"` in [my original version](https://github.com/JuliaLang/julia/pull/34111/commits/cba9a8a561a004c098de7ec3a9fd2717296016ab) of  `escape_raw_string` was that I really couldn't think of any reason not to: the entire function really only makes sense if its output is immediately followed by a `"`, because of the way this encoding specially treats trailing backslashes at the end of the string (expecting the recipient to see that followed by a `"`). This is different from how `escape_string()` works, where not only the delimiter is configurable, but also nothing special happens to backslashes at the end of the string. I'd therefore stay with my [original version](https://github.com/JuliaLang/julia/pull/34111/commits/cba9a8a561a004c098de7ec3a9fd2717296016ab), as the analogy to `escape_string()` is not so clear cut.
(The difference in mathematical terms: `escape_string()` is an isomorphism over string concatenation, whereas `escape_string_raw()` isn't, because of the case of strings ending in a backslash.)

**Comment by JeffBezanson** [comment]
That makes sense, but it's easy to add the quotes if you want them, and impossible to remove them if you don't (however unlikely that might be).

File: base/strings/io.jl [hunk 1]
```diff
@@ -523,6 +523,55 @@ julia> println(raw"\\\\x \\\\\\"")
...
+                # if one or more backslashes are followed by
+                # a double quote then escape all backslashes
+                # and the double quote
+                escapes = escapes * 2 + 1
+            end
+            while escapes > 0
+                write(io, '\\')
+                escapes -= 1
+            end
+            escapes = 0
```
Comment (musm) [inline base/strings/io.jl:560]: Looks like this line is unecessary since  escape is decremented to 0 by L559

Comment (mgkuhn) [reply base/strings/io.jl:560]: Agreed.

**Pull request merged**